Calculator: shunting yard, added exponentiation and parentheses

EmmanuelMess opened the pull request:
The functionality is more or less the same as before but:

* Fixed underflow? (fixed `1 + 1.11111 = 1.1111112`)
* Added exponentiation (`-1^2 = -1` or `-1**2 = -1`)
* Added parentheses (`(1+1) = 2`)
* Added unary + (`+1*2 = 2`)

## Help needed
* To find operations that break this implementation, be it a crash or incorrect answer.
* To fix the current RegEx so that any expression can be found.
* To find expressions that every iteration of RegEx can't parse correctly (find false positives or false negatives).

Fixes #1217
Fixes point 3 of #1151
Fixed #1221

----
<a class="imgpatreon" href="https://www.patreon.com/emmanuelmess" target="_blank">
<img alt="Become a patreon" src="https://user-images.githubusercontent.com/10991116/56376378-07065400-61de-11e9-9583-8ff2148aa41c.png" width=150px></a>

EmmanuelMess commented:
The current RegEx is:
```^(-?)([0-9.]+)\s?([+\-*/×x÷^])\s?(-?)([0-9.]+)$```
It breaks with '1+1..' and every operation that has more than one operator.
And a possible replacement would be:
```^([-+]? ?(\d+|\(\g<1>\))( ?[-+*\/] ?\g<1>)?)$```
but named groups are not supported.

Neamar commented on app/src/main/java/fr/neamar/kiss/utils/calculator/Calculator.java:
```diff
@@ -0,0 +1,62 @@
+				case Tokenizer.Token.EXP_TOKEN:
+					operand2 = stack.pop();
+					operand1 = stack.pop();
+					stack.push(operand1.pow(operand2.intValueExact()));
```
What if the operand is not an int?

EmmanuelMess replied:
Forgot to code half of that, thanks!

Neamar commented:
I gave it a go! It's pretty nice :)

A couple issues:

* division by zero crashes the app
* percentages are not handled, which is fine, but still trigger the regexp. So 50 + 20% is 70 :( I'm fine if we just don't consider the expression as valid
* same with `!`, the user may expect factorial something. Same with € or $ -- we may want to just drop support for anything that isn't a number or an operator.
* regexp is unbounded, which I think is an issue -- I can search for "qsklndl1+1flm," and get 2
* very large operations crash KISS (anything that seems to reach Infinity or NaN) [e.g. `99^999`]
* Multiple decimal separator will crash (e.g. "8.8.8")
* unary operators can be chained. I'm not sure what I expect to see, but "4-+4" shouldn't make sense (I guess the opposite is fine, 4+-4 is OK)
* using spaces break the code. In France for instance, thousand separator is a space " ": "1 000" is one thousand. However "1 000 * 5" is 0 (displayed as 000 * 5). I don't think we need to parse this, but we shouldn't return invalid results -- back to the point about the regexp.

Looking at all this, I think it's unlikely we'll get a regexp that matches anything that is syntactically correct.

We may want to just use a very generic regexp (`^[-.0-9+*/^]+$`) and then, while parsing, if we get an expression that doesn't make sense, we just don't return anything from the CalculatorProvider? (with some special cases, for instance `2+1.` should display 3 since we can assume the user will enter the first decimal after)

EmmanuelMess commented:
>     * division by zero crashes the app

```
2019-06-20 12:26:47.834 12318-12318/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: fr.neamar.kiss, PID: 12318
    java.lang.IndexOutOfBoundsException: setSpan (-1 ... 19) starts before 0
        at android.text.SpannableStringInternal.checkRange(SpannableStringInternal.java:496)
        at android.text.SpannableStringInternal.setSpan(SpannableStringInternal.java:196)
        at android.text.SpannableStringInternal.setSpan(SpannableStringInternal.java:183)
        at android.text.SpannableString.setSpan(SpannableString.java:60)
        at fr.neamar.kiss.result.Result.displayHighlighted(Result.java:96)
        at fr.neamar.kiss.result.SearchResult.display(SearchResult.java:62)
        at fr.neamar.kiss.adapter.RecordAdapter.getView(RecordAdapter.java:110)
        at android.widget.AbsListView.obtainView(AbsListView.java:2405)
```

wut? I'll skip printing "ARITHMETIC ERROR" and just return.


>     * percentages are not handled, which is fine, but still trigger the regexp. So 50 + 20% is 70 :( I'm fine if we just don't consider the expression as valid
> 
>     * same with `!`, the user may expect factorial something. Same with € or $ -- we may want to just drop support for anything that isn't a number or an operator.
> 
>     * regexp is unbounded, which I think is an issue -- I can search for "qsklndl1+1flm," and get 2

I thought it would be a good idea to make the RegEx unbounded, it isn't. Fixed.

>     * very large operations crash KISS (anything that seems to reach Infinity or NaN) [e.g. `99^999`]

Fixed.

>     * Multiple decimal separator will crash (e.g. "8.8.8")

Fixed.

>     * unary operators can be chained. I'm not sure what I expect to see, but "4-+4" shouldn't make sense (I guess the opposite is fine, 4+-4 is OK)

Unary operators are operators and can be used on any operand. What do you suggest?

>     * using spaces break the code. In France for instance, thousand separator is a space " ": "1 000" is one thousand. However "1 000 * 5" is 0 (displayed as 000 * 5). I don't think we need to parse this, but we shouldn't return invalid results -- back to the point about the regexp.

Ok, fixed, but a side effect is that in Argentinian locale "1.1" gets converted to "11" because '.' is the thousand separator. Also fixed the output side of this.

>     * a weird issue with floats? Not sure what's happening, but try `4*(2+5/4`: this returns `13.00`, why not just 13?

BigDecimal tries to save as much precision as possible, fixed with the localizer at the output.

>     * unmatched closing paren can kill the app: `5+2)`

Ups, fixed.

> Looking at all this, I think it's unlikely we'll get a regexp that matches anything that is syntactically correct.

Impossible I would say, the android dialect doesn't allow for recursive regexes.

> We may want to just use a very generic regexp (`^[-.0-9+*/^]+$`) and then, while parsing, if we get an expression that doesn't make sense, we just don't return anything from the CalculatorProvider? (with some special cases, for instance `2+1.` should display 3 since we can assume the user will enter the first decimal after)

That's what I tried to make, I've remade the parser to allow it to return error states, and the RegEx is just checking for the start and the end of an operation, and checking that there are no alphabetical characters in between.
And it is broken for thosand separators

EmmanuelMess commented:
Another solution for the RegEx is using Java for 'augmenting' the capabilities of the RegEx dialect, or skipping the RegEx completely and creating an automata for parsing mathematical expressions.

Neamar commented:
Hello!

Thanks for the update :) I tried some other things...

* Typing "2**" crashes the app
* Actually... typing "2*" crashes the app! I think that's a bug ;)
* 8.8.8 + 1 returns a result (9.8). I find it weird, not sure how you feel about it?
* Division by zero:

EmmanuelMess commented:
>     * Typing "2**" crashes the app
>
>     * Actually... typing "2*" crashes the app! I think that's a bug ;)

Fixed!
 
>     * 8.8.8 + 1 returns a result (9.8). I find it weird, not sure how you feel about it?

Java's `DecimalFormat` takes '8.8.8' and doesn't throw an exception, fixing this would be hell because we have to check for localization and only then check for multiple decimal separators 

>     * Division by zero:

"2/0" doesn't crash. If you want me to implement limits I will, not in this PR though.

Neamar commented:
Nice!

Just one more thing, do you think you can get the tests to run in the CI? I don't think that's the case for now.

Thanks, after that I think we can merge!

EmmanuelMess commented:
> Just one more thing, do you think you can get the tests to run in the CI? I don't think that's the case for now.

Done!

ildar commented:
Guys, so many special cases make me think about unit testing..

EmmanuelMess commented:
> Guys, so many special cases make me think about unit testing..

If you want me to add cases to the unit tests I have no problems, just list them.

Neamar commented:
@EmmanuelMess I think @ildar didn't notice them :)

I don't have access to my personal computer this week, so I'll test one last time next week and merge! But this looks all good for me.

ildar commented:
Oh! Sorry

Neamar commented on app/src/main/java/fr/neamar/kiss/dataprovider/simpleprovider/CalculatorProvider.java:
```diff
@@ -1,73 +1,70 @@
+            String queryProcessed = operation + readableResult;
             SearchPojo pojo = new SearchPojo("calculator://", queryProcessed, "", SearchPojo.CALCULATOR_QUERY);
 
             pojo.relevance = 100;
```
That's not from you, but I believe this should now be 19. Otherwise, when you type a phone number, the math result appears before the call option

EmmanuelMess replied:
The relevance? I have no idea what that means, and it is not related to this PR.

Neamar replied:
It's the importance of a result -- how high in the list it should be. It's fine if you don't change it, I'll commit something after merging

Neamar commented:
I tried it again! Two last comments:

* one about relevant for CalculatorPojo, this should be 19 -- see comment on code
* I can't get negative powers (e.g. `2^-1`), or `2^+1`. It's not a huge deal, since using parenthesis works.

Neamar commented on app/src/main/java/fr/neamar/kiss/dataprovider/simpleprovider/CalculatorProvider.java:
```diff
@@ -1,73 +1,70 @@
     public void requestResults(String query, Searcher searcher) {
-        // Now create matcher object.
-        Matcher m = p.matcher(query);
+        // Now create matcher objct.
```
object ;)

Neamar commented:
Thanks for your contribution :)

Neamar commented:
One thing that could be nice if you have time: if there is no operator, skip returning. For instance if I type "0669347018" which is a phone number, I get the math result that just redisplays my input. Might be worth a second small PR if you can :)

Neamar commented:
(I think I merged slightly too early, sorry. Feel free to send another PR if you want!)

EmmanuelMess commented:
> (I think I merged slightly too early, sorry. Feel free to send another PR if you want!)

Don't worry, I'll add some issues and get to it later.
